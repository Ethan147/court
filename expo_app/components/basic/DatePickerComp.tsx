import React, { useState } from "react";
import { Button, Platform, View, StyleProp, ViewStyle } from "react-native";
import DateTimePickerModal from "react-native-modal-datetime-picker";

import TextInputComp from "./TextInputComp";

interface DatePickerCompProps {
  label: string;
  webFormatHint: string;
  value: string;
  onDateChange: (text: string) => void;
  onBlur?: () => void;
  error?: boolean;
  passStyles?: {
    datePickerCompWebView?: StyleProp<ViewStyle>;
    datePickerCompWebInput?: StyleProp<ViewStyle>;
    datePickerCompAppView?: StyleProp<ViewStyle>;
    datePickerCompAppModal?: StyleProp<ViewStyle>;
    datePickerCompAppButton?: StyleProp<ViewStyle>;
    // text input
    textInputCompOuterView?: StyleProp<ViewStyle>;
    textInputCompTextContainer?: StyleProp<ViewStyle>;
    textInputCompIconContainer?: StyleProp<ViewStyle>;
    textInputCompViewStyle?: StyleProp<ViewStyle>;
    textInputCompText?: StyleProp<ViewStyle>;
    textInputCompHintMessage?: StyleProp<ViewStyle>;
  };
}

const DatePickerComp: React.FC<DatePickerCompProps> = ({
  label,
  webFormatHint,
  value,
  onDateChange,
  onBlur,
  error,
  passStyles,
}) => {
  const styles = {
    datePickerCompWebView: passStyles?.datePickerCompWebView || {},
    datePickerCompWebInput: passStyles?.datePickerCompWebInput || {},
    datePickerCompAppView: passStyles?.datePickerCompAppView || {},
    datePickerCompAppModal: passStyles?.datePickerCompAppModal || {},
    datePickerCompAppButton: passStyles?.datePickerCompAppButton || {},
    // text input for web
    textInputCompOuterView: passStyles?.textInputCompOuterView || {},
    textInputCompTextContainer: passStyles?.textInputCompTextContainer || {},
    textInputCompIconContainer: passStyles?.textInputCompIconContainer || {},
    textInputCompViewStyle: passStyles?.textInputCompViewStyle || {},
    textInputCompText: passStyles?.textInputCompText || {},
    textInputCompHintMessage: passStyles?.textInputCompHintMessage || {},
  };

  const [dateValue, setDateValue] = useState<string>("");
  const [isDatePickerVisible, setDatePickerVisibility] =
    useState<boolean>(false);

  const showDatePicker = () => {
    setDatePickerVisibility(true);
  };

  const hideDatePicker = () => {
    setDatePickerVisibility(false);
  };

  const handleBlur = () => {
    if (onBlur) {
      onBlur();
    }
  };

  const handleConfirm = (date: Date) => {
    console.warn("A date has been picked: ", date);
    const formattedDate = date.toLocaleDateString();
    setDateValue(formattedDate);
    onDateChange(formattedDate);
    hideDatePicker();
    handleBlur();
  };

  const handleWebDateChange = (text: string) => {
    // remove all non-numeric characters
    let newText = text.replace(/[^0-9]/g, "");

    // add slashes at the correct positions
    if (newText.length >= 2) {
      newText = newText.slice(0, 2) + "/" + newText.slice(2);
    }
    if (newText.length >= 5) {
      newText = newText.slice(0, 5) + "/" + newText.slice(5);
    }

    onDateChange(newText);
    console.warn("A date has been picked: ", newText);
    handleBlur();
  };

  if (Platform.OS === "web") {
    return (
      <View style={styles.datePickerCompWebView}>
        <TextInputComp
          label={label + " " + webFormatHint}
          value={value}
          error={error}
          onChangeText={handleWebDateChange}
          passStyles={styles}
        />
      </View>
    );
  }

  const buttonText = dateValue ? dateValue : "select " + label;

  return (
    <View style={styles.datePickerCompAppView}>
      <Button title={buttonText} onPress={showDatePicker} />
      <DateTimePickerModal
        isVisible={isDatePickerVisible}
        mode="date"
        onConfirm={handleConfirm}
        onCancel={hideDatePicker}
        style={styles.datePickerCompAppModal}
      />
    </View>
  );
};

export default DatePickerComp;
