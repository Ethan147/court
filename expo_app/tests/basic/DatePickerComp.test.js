import { render, fireEvent } from "@testing-library/react-native";
import React from "react";
import { Platform } from "react-native";
import DateTimePickerModal from "react-native-modal-datetime-picker";

import DatePickerComp from "../../components/basic/DatePickerComp";

jest.mock("react-native-modal-datetime-picker");

describe("<DatePickerComp />", () => {
  const onDateChangeMock = jest.fn();
  const onBlurMock = jest.fn();

  const props = {
    label: "Test Date",
    webFormatHint: "(mm/dd/yyyy)",
    value: "01/01/2022",
    onDateChange: onDateChangeMock,
    onBlur: onBlurMock,
    error: false,
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const originalError = console.error;

    console.error = jest.fn(); // mock console.error
    render(<DatePickerComp {...props} />);
    console.error = originalError; // restore console.error
  });

  /*
  describe('Web Platform', () => {
    beforeAll(() => {
      Platform.OS = 'web';
    });

    it('displays the passed value', () => {
      const { getByDisplayValue } = render(<DatePickerComp {...props} />);
      expect(getByDisplayValue('01/01/2022')).toBeDefined();
    });

    it('calls onDateChange with the correct formatted date', () => {
      const { getByDisplayValue } = render(<DatePickerComp {...props} />);
      const input = getByDisplayValue('01/01/2022');
      fireEvent.changeText(input, '20012003');
      expect(onDateChangeMock).toHaveBeenCalledWith('20/01/2003');
    });

    // More web tests go here...
  });
  */

  /*
  describe('Non-web Platforms', () => {
    beforeAll(() => {
      Platform.OS = 'ios'; // or 'android'
    });

    it('opens DateTimePickerModal on button press', () => {
      const { getByText } = render(<DatePickerComp {...props} />);
      const button = getByText('select Test Date');
      fireEvent.press(button);
      expect(DateTimePickerModal).toBeVisible();
    });

    it('calls handleConfirm when date is selected', () => {
      const { getByText } = render(<DatePickerComp {...props} />);
      const button = getByText('select Test Date');
      fireEvent.press(button);
      // you can mock the confirm action of the DateTimePickerModal here
      // but as of the last update, there's no official way to do it using '@testing-library/react-native'
    });

    // More non-web tests go here...
  });
  */
});
