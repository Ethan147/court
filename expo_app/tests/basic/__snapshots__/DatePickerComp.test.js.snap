// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePickerComp /> snapshot 1`] = `
<View
  style={{}}
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="datePickerButton"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        select Test Date
      </Text>
    </View>
  </View>
  <Modal
    animationType="none"
    hardwareAccelerated={false}
    isVisible={false}
    onHide={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
